### PR TITLE
[RDY]"BrightNights" - the sci-fi mod

### PR DESCRIPTION
**Pull request by Coolthulhu**
# Rationale

DDA is supposed to have more or less realistic limitations. No craftable laser guns , no psionic powers, cryo-grenades, forcefields, not even automatic weapons crafted in garage.
Some limitations outlined in http://smf.cataclysmdda.com/index.php?topic=10444.0
Plus, DDA is pretty simulationistic, with a lot of items that exist only because they do in real life.

So here's a mod that is supposed to contain those unrealistic elements, but (later on) not the realistic ones.

Additionally, it is supposed to be:
* "Example mod" for new features I'd like to show off
* A way to provide more Lua examples, hopefully to get more people to use it. Unless Lua gets axed, see #20172
* A way to test the mod system to find what it needs
* (Eventually) A light core mod, with less features than the full DDA

___

# Actually implemented

In this PR I added a mod that has:
* Hand-craftable laser gun
* Redefined recharging station - from a complex military system to common battery charger, craftable by mediocre electrician
* Rechargable batteries that fit the two above

Also two code changes:
* Enabled vehicle charging stations to charge non-tools. In this case, the batteries.
* In charging code, rearranged sides of `&&` to start with the more restrictive and faster to execute check first

___

# QA

> Mod X is not in mainline, why this one should be?

Outlined in #20129
tl;dr Non-huge mod maintained in core repo, by an active maintainer, is unlikely to cause trouble.
Plus, this mod is supposed to serve as an example mod for some future new features.

> Item X is sci-fi. Does it go here?

Maybe. This mod isn't intended to be a dump for sci-fi items rejected from DDA, but could end up "rescuing" some of them, provided they have a gameplay reason to be here.

> Can I contribute to the mod?

Yes, but mind the design rules. Those aren't written yet, but will be, if this PR goes smoothly.

___

# [CR]

[CR] because I want to resolve any potential conflicts - personal or otherwise.

Also because I'm not sure about the category of the mod - I'd like to make it a core mod eventually, but at the moment it is an item addition mod. If it is supposed to eventually conflict with DDA mod, should it be classified as core mod or as addition mod?

### CODE REVIEW

**Comment by kevingranade** [comment]
I need to double check what "core" actually means, I think it was adjusted
during my hiatus with the creation of the "dda" mod, so I'm not clear on
whether it's supposed to be "essentially part of the game engine" or "a
core component intended for extension".

Most likely "core" is the right place, when selecting mods for a new world
a player would deselect "dda", select "BrightNights" (this might just mean
selecting "BrightNights", which would toggle "dda" off), then select mods
that extend "BrightNights" in subsequent tabs.

**Comment by Coolthulhu** [comment]
OK, moved to core category, which is supposed to eventually actually be a part of.

Most of the additions in this specific PR could easily be mainlined, but for now I'll keep them in the mod since most of the tools in the game do not support using rechargable batteries without a mod.

**Comment by keyspace** [comment]
IMO PR OK.

-----

> I need to double check what "core" actually means, I think it was adjusted during my hiatus with the creation of the "dda" mod, so I'm not clear on whether it's supposed to be "essentially part of the game engine" or "a core component intended for extension". 

More the latter, but _both_ are present.

For ref: PR #19260 on core `dda`, PR #19054 for `modinfo` fields.

So:

* `data/core` - "always present",
* `dda` - "core content pack" that makes a certain kind of game.

Hasn't been documented much outside those PRs. The former notes:

> Next step is to move the minimal number of items from dda to data/core and then other contributors can begin to write total conversion mods.

As things stand, the always-present `data/core` only has `water` (`ammunition_type` and `COMESTIBLE`), and `dda` mod's files are familiarly spread all over `data/*`.

**Comment by keyspace** [comment]
Actually, haven't tested what happens with making this PR's mod a `content` pack.

**Comment by keyspace** [comment]
> If it is supposed to eventually conflict with DDA mod, should it be classified as core mod or as addition mod?

AFAIU from PR #19260, "core content packs" are supposed to conflict with each other. So classifying this as `content` is correct (if it's intended to conflict later on), but having a dependency on `dda` is not.

This is all moot, though, since there are no live examples, and migration of "core items" from `dda` to `data/core` hasn't been undertaken.

**Comment by keyspace** [comment]
I was thinking more of what this would do to a save that was started with `BrightNights` having a dependency on `dda`, but then loaded on a build where it was no longer the case, and perhaps not having more than one "core content pack" was enforced.

This is very far into the future, maybe no sense to think about it now. But perhaps reclassifying it into a "core content pack" is best left for when there's an actual conflict?..

Anyway, these are idle musings.
